src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace_SolutionCaching.cs

```diff
@@ -0,0 +1,182 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    internal sealed partial class RemoteWorkspace
+    {
+        /// <summary>
+        /// The last solution requested by a service. This effectively adds an additional in-flight-count to one of
+        /// the items in <see cref="_solutionChecksumToSolution"/> ensuring that the very last solution requested is
+        /// kept alive by us, even if there are no active requests currently in progress for that solution.  That
+        /// way if we have two non-concurrent requests for that same solution, with no intervening updates, we can
+        /// cache and keep the solution around instead of having to recompute it.
+        /// </summary>
+        private InFlightSolution? _lastAnyBranchSolution;
+
+        /// <summary>
+        /// The last solution requested by a service. This effectively adds an additional in-flight-count to one of
+        /// the items in <see cref="_solutionChecksumToSolution"/> ensuring that the very last solution requested is
+        /// kept alive by us, even if there are no active requests currently in progress for that solution.  That
+        /// way if we have two non-concurrent requests for that same solution, with no intervening updates, we can
+        /// cache and keep the solution around instead of having to recompute it.
+        /// </summary>
+        private InFlightSolution? _lastPrimaryBranchSolution;
+
+        /// <summary>
+        /// Mapping from solution-checksum to the solution computed for it.  This is used so that we can hold a
+        /// solution around as long as the checksum for it is being used in service of some feature operation (e.g.
+        /// classification).  As long as we're holding onto it, concurrent feature requests for the same solution
+        /// checksum can share the computation of that particular solution and avoid duplicated concurrent work.
+        /// </summary>
+        private readonly Dictionary<Checksum, InFlightSolution> _solutionChecksumToSolution = new();
+
+        private async ValueTask<InFlightSolution> GetOrCreateSolutionAndAddInFlightCountAsync(
+            Checksum solutionChecksum,
+            Func<CancellationToken, Task<Solution>> computeDisconnectedSolutionAsync,
+            Func<Solution, CancellationToken, Task<Solution>>? updatePrimaryBranchAsync,
+            CancellationToken cancellationToken)
+        {
```

Comment:
@dibarbet this takes your suggestion of just having a single method that atomically does all the right things wrt to setting up state.